File: js/lib/LocationBuilder.ts
import { Location, LocationGroup, Locations, isDigitCode } from "./Location";
import { VisVersion } from "./VisVersion";

export class LocationBuilder {
    public number?: number;
    public side?: string;
    public vertical?: string;
    public transverse?: string;
    public longitudinal?: string;

    public visVersion: VisVersion;

    protected _reversedGroups: Map<string, LocationGroup>;

    private constructor(previous: LocationBuilder);
    private constructor(
        visVersion: VisVersion,
        reversedGroups: Map<string, LocationGroup>
    );
    private constructor(
        arg1: VisVersion | LocationBuilder,
        arg2?: Map<string, LocationGroup>
    ) {
        if (arg1 instanceof LocationBuilder && arg2 === undefined) {
            this.number = arg1.number;
            this.side = arg1.side;
            this.vertical = arg1.vertical;
            this.transverse = arg1.transverse;
            this.longitudinal = arg1.longitudinal;
            this.visVersion = arg1.visVersion;
            this._reversedGroups = arg1._reversedGroups;
        } else {
            this.visVersion = arg1 as VisVersion;
            this._reversedGroups = arg2!;
        }
    }

    public static create(locations: Locations) {
        return new LocationBuilder(
            locations.visVersion,
            locations.reversedGroups
        );
    }

    public get(group: LocationGroup): number | string | undefined {
        return {
            [LocationGroup.Number]: this.number,
            [LocationGroup.Side]: this.side,
            [LocationGroup.Vertical]: this.vertical,
            [LocationGroup.Transverse]: this.transverse,
            [LocationGroup.Longitudinal]: this.longitudinal,
        }[group];
    }

    public tryWithLocation(value?: Location): LocationBuilder {
        if (!value) return this;
        return this.withLocation(value);
    }

    public withLocation(value: Location): LocationBuilder {
        let builder = this.clone();

        const span = value.toString();

        let n: number | undefined = undefined;

        for (let i = 0; i < span.length; i++) {
            const ch = span.charAt(i);

            if (isDigitCode(ch)) {
                if (n === undefined) n = +ch;
                else {
                    n = parseInt(span.slice(0, i + 1));
                    if (isNaN(n))
                        throw new Error("Should include a valid number");
                }
                continue;
            }
            builder = builder.withValue(ch);
        }

        if (n !== undefined) builder = builder.withNumber(n);

        return builder;
    }

    public withNumber(value: number): LocationBuilder {
        return this.withValueInternal(LocationGroup.Number, value);
    }

    public withSide(value: string): LocationBuilder {
        return this.withValueInternal(LocationGroup.Side, value);
    }

    public withVertical(value: string): LocationBuilder {
        return this.withValueInternal(LocationGroup.Vertical, value);
    }

    public withTransverse(value: string): LocationBuilder {
        return this.withValueInternal(LocationGroup.Transverse, value);
    }

    public withLongitudinal(value: string): LocationBuilder {
        return this.withValueInternal(LocationGroup.Longitudinal, value);
    }

    public withValue(
        value: number | string,
        group: LocationGroup | undefined = undefined
    ): LocationBuilder {
        if (group !== undefined) return this.withValueInternal(group, value);
        if (typeof value === "number") {
            return this.withValueInternal(LocationGroup.Number, value);
        } else {
            const key = this._reversedGroups.get(value);
            if (!key)
                throw new Error(
                    `The value ${value} is an invalid Locations value`
                );
            return this.withValueInternal(key, value);
        }
    }

    private withValueInternal(
        group: LocationGroup,
        value: number | string
    ): LocationBuilder {
        if (group === LocationGroup.Number) {
            if (typeof value !== "number")
                throw new Error("Value should be number");
            if (value < 1) throw new Error("Value should be greater than 0");
            if (value % 1 != 0)
                throw new Error("Value must be an integer larger than 0");

            return this.with((s) => (s.number = value));
        }

        if (typeof value !== "string")
            throw new Error("Value should be a string");
        if (value.length !== 1)
            throw new Error("Value should be a single character");

        const key = this._reversedGroups.get(value);
        if (key === undefined || key !== group) {
            throw new Error(
                `The value ${value} is an invalid ${LocationGroup[group]} value`
            );
        }

        return {
            [LocationGroup.Side]: this.with((s) => (s.side = value)),
            [LocationGroup.Vertical]: this.with((s) => (s.vertical = value)),
            [LocationGroup.Transverse]: this.with(
                (s) => (s.transverse = value)
            ),
            [LocationGroup.Longitudinal]: this.with(
                (s) => (s.longitudinal = value)
            ),
        }[key];
    }

    public withoutNumber(): LocationBuilder {
        return this.withoutValue(LocationGroup.Number);
    }

    public withoutSide(): LocationBuilder {
        return this.withoutValue(LocationGroup.Side);
    }

    public withoutVertical(): LocationBuilder {
        return this.withoutValue(LocationGroup.Vertical);
    }

    public withoutTransverse(): LocationBuilder {
        return this.withoutValue(LocationGroup.Transverse);
    }

    public withoutLongitudinal(): LocationBuilder {
        return this.withoutValue(LocationGroup.Longitudinal);
    }

    public withoutValue(group: LocationGroup) {
        return {
            [LocationGroup.Number]: this.with((s) => (s.number = undefined)),
            [LocationGroup.Side]: this.with((s) => (s.side = undefined)),
            [LocationGroup.Vertical]: this.with(
                (s) => (s.vertical = undefined)
            ),
            [LocationGroup.Transverse]: this.with(
                (s) => (s.transverse = undefined)
            ),
            [LocationGroup.Longitudinal]: this.with(
                (s) => (s.longitudinal = undefined)
            ),
        }[group];
    }

    public with(u: { (state: LocationBuilder): void }): LocationBuilder {
        const n = this.clone();
        u && u(n);
        return n;
    }

    public clone(): LocationBuilder {
        return new LocationBuilder(this);
    }

    public build(): Location {
        return new Location(this.toString());
    }

    public toString(): string {
        const items = [
            this.side,
            this.vertical,
            this.transverse,
            this.longitudinal,
        ];

        let s = "";

        for (const item of items) {
            if (item === undefined) continue;
            s += item;
        }

        const chars = s.split("").sort().join("");

        return this.number !== undefined ? `${this.number}${chars}` : chars;
    }
}

File: cpp/src/dnv/vista/sdk/LocationBuilder.cpp
/**
 * @file LocationBuilder.cpp
 * @brief Implementation of the LocationBuilder class
 */

#include "pch.h"

#include "dnv/vista/sdk/LocationBuilder.h"

#include "dnv/vista/sdk/Locations.h"
#include "dnv/vista/sdk/VISVersion.h"
#include "dnv/vista/sdk/utils/StringBuilderPool.h"
#include "dnv/vista/sdk/Exceptions.h"

namespace dnv::vista::sdk
{
	//=====================================================================
	// LocationBuilder class
	//=====================================================================

	//----------------------------------------------
	// Construction
	//----------------------------------------------

	LocationBuilder::LocationBuilder( VisVersion visVersion, const std::map<char, LocationGroup>* reversedGroups )
		: m_visVersion{ visVersion },
		  m_reversedGroups{ reversedGroups }
	{
	}

	//----------------------------------------------
	// Conversion and comparison
	//----------------------------------------------

	std::string LocationBuilder::toString() const
	{
		auto lease = utils::StringBuilderPool::instance();
		auto builder = lease.Builder();

		if ( m_number.has_value() )
		{
			builder.append( std::to_string( m_number.value() ) );
		}

		if ( m_side.has_value() )
		{
			builder.push_back( m_side.value() );
		}
		if ( m_vertical.has_value() )
		{
			builder.push_back( m_vertical.value() );
		}
		if ( m_transverse.has_value() )
		{
			builder.push_back( m_transverse.value() );
		}
		if ( m_longitudinal.has_value() )
		{
			builder.push_back( m_longitudinal.value() );
		}

		std::string result = lease.toString();
		std::sort( result.begin(), result.end() );

		return result;
	}

	//----------------------------------------------
	// Static factory methods
	//----------------------------------------------

	LocationBuilder LocationBuilder::create( const Locations& locations )
	{
		return LocationBuilder( locations.visVersion(), &locations.reversedGroups() );
	}

	//----------------------------------------------
	// Build methods (Immutable fluent interface)
	//----------------------------------------------

	//----------------------------
	// Build
	//----------------------------

	Location LocationBuilder::build() const
	{
		return Location( toString() );
	}

	//----------------------------
	// Location
	//----------------------------

	LocationBuilder LocationBuilder::withLocation( const Location& location ) const
	{
		LocationBuilder builder{ m_visVersion, m_reversedGroups };
		
		std::string locationStr = location.toString();
		std::string_view span = locationStr;
		std::optional<int> number;

		for ( size_t i = 0; i < span.length(); ++i )
		{
			char ch = span[i];

			if ( std::isdigit( ch ) )
			{
				if ( !number.has_value() )
				{
					number = ch - '0';
				}
				else
				{
					int parsedNumber;
					if ( !Locations::tryParseInt( span, 0, static_cast<int>( i + 1 ), parsedNumber ) )
					{
						throw std::invalid_argument( "Should include a valid number" );
					}
					number = parsedNumber;
				}
				continue;
			}

			builder = builder.withValue( ch );
		}

		if ( number.has_value() )
		{
			builder = builder.withNumber( number.value() );
		}

		return builder;
	}

	//----------------------------
	// Number
	//----------------------------

	LocationBuilder LocationBuilder::withNumber( int number ) const
	{
		return withValueInternal( LocationGroup::Number, number );
	}

	LocationBuilder LocationBuilder::withoutNumber() const
	{
		return withoutValue( LocationGroup::Number );
	}

	//----------------------------
	// Side
	//----------------------------

	LocationBuilder LocationBuilder::withSide( char side ) const
	{
		return withValueInternal( LocationGroup::Side, side );
	}

	LocationBuilder LocationBuilder::withoutSide() const
	{
		return withoutValue( LocationGroup::Side );
	}

	//----------------------------
	// Vertical
	//----------------------------

	LocationBuilder LocationBuilder::withVertical( char vertical ) const
	{
		return withValueInternal( LocationGroup::Vertical, vertical );
	}

	LocationBuilder LocationBuilder::withoutVertical() const
	{
		return withoutValue( LocationGroup::Vertical );
	}

	//----------------------------
	// Transverse
	//----------------------------

	LocationBuilder LocationBuilder::withTransverse( char transverse ) const
	{
		return withValueInternal( LocationGroup::Transverse, transverse );
	}

	LocationBuilder LocationBuilder::withoutTransverse() const
	{
		return withoutValue( LocationGroup::Transverse );
	}

	//----------------------------
	// Longitudinal
	//----------------------------

	LocationBuilder LocationBuilder::withLongitudinal( char longitudinal ) const
	{
		return withValueInternal( LocationGroup::Longitudinal, longitudinal );
	}

	LocationBuilder LocationBuilder::withoutLongitudinal() const
	{
		return withoutValue( LocationGroup::Longitudinal );
	}

	//----------------------------
	// Internal implementation
	//----------------------------

	LocationBuilder LocationBuilder::withValueInternal( LocationGroup group, int value ) const
	{
		if ( group != LocationGroup::Number )
		{
			throw ValidationException( "Integer values are only valid for Number group" );
		}

		if ( value < 1 )
		{
			throw ValidationException( "Value should be greater than 0" );
		}

		LocationBuilder result = *this;
		result.m_number = value;

		return result;
	}

	LocationBuilder LocationBuilder::withValueInternal( LocationGroup group, char value ) const
	{
		if ( group == LocationGroup::Number )
		{
			throw std::invalid_argument( "Character values are not valid for Number group" );
		}

		const auto it = m_reversedGroups->find( value );
		if ( it == m_reversedGroups->end() || it->second != group )
		{
			const char* groupName = [group]() {
				switch ( group )
				{
					case LocationGroup::Side:
					{
						return "Side";
					}
					case LocationGroup::Vertical:
					{
						return "Vertical";
					}
					case LocationGroup::Transverse:
					{
						return "Transverse";
					}
					case LocationGroup::Longitudinal:
					{
						return "Longitudinal";
					}
					case LocationGroup::Number:
					{
						return "Number";
					}
					default:
					{
						return "Unknown";
					}
				}
			}();

			throw ValidationException(
				fmt::format( "The value '{}' is an invalid {} value", value, groupName ) );
		}

		LocationBuilder result = *this;
		switch ( group )
		{
			case LocationGroup::Side:
			{
				result.m_side = value;
				break;
			}
			case LocationGroup::Vertical:
			{
				result.m_vertical = value;
				break;
			}
			case LocationGroup::Transverse:
			{
				result.m_transverse = value;
				break;
			}
			case LocationGroup::Longitudinal:
			{
				result.m_longitudinal = value;
				break;
			}
			case LocationGroup::Number:
			{
				throw std::invalid_argument( "Number group should not contain character values" );
			}
			default:
			{
				throw std::invalid_argument( "Unsupported LocationGroup" );
			}
		}
		return result;
	}

	//----------------------------
	// Value
	//----------------------------

	LocationBuilder LocationBuilder::withValue( int value ) const
	{
		return withValueInternal( LocationGroup::Number, value );
	}

	LocationBuilder LocationBuilder::withValue( char value ) const
	{
		const auto it = m_reversedGroups->find( value );
		if ( it == m_reversedGroups->end() )
		{
			throw ValidationException( fmt::format( "The value '{}' is an invalid Locations value", value ) );
		}

		return withValueInternal( it->second, value );
	}

	LocationBuilder LocationBuilder::withoutValue( LocationGroup group ) const
	{
		switch ( group )
		{
			case LocationGroup::Number:
			{
				return withoutNumber();
			}
			case LocationGroup::Side:
			{
				return withoutSide();
			}
			case LocationGroup::Vertical:
			{
				return withoutVertical();
			}
			case LocationGroup::Transverse:
			{
				return withoutTransverse();
			}
			case LocationGroup::Longitudinal:
			{
				return withoutLongitudinal();
			}
			default:
			{
				throw std::invalid_argument( "Unsupported LocationGroup" );
			}
		}
	}
}
